File: e2e-tests/spring-examples/src/test/java/org/evomaster/e2etests/spring/examples/db/base/DbBaseEMTest.java
package org.evomaster.e2etests.spring.examples.db.base;

import org.evomaster.core.problem.rest.HttpVerb;
import org.evomaster.core.problem.rest.RestIndividual;
import org.evomaster.core.search.Solution;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class DbBaseEMTest extends DbBaseTestBase {


    @Test
    public void testRunEM() throws Throwable {

        runTestHandlingFlakyAndCompilation(
                "DbBaseEM",
                "org.bar.db.BaseEM",
                10_000,
                (args) -> {

                    args.add("--heuristicsForSQL");
                    args.add("true");
                    args.add("--generateSqlDataWithSearch");
                    args.add("false");

                    //FIXME: need to study and fix the side effects of Taint here
                    args.add("--baseTaintAnalysisProbability");
                    args.add("0.0");

                    Solution<RestIndividual> solution = initAndRun(args);

                    assertTrue(solution.getIndividuals().size() >= 1);

                    assertHasAtLeastOne(solution, HttpVerb.GET, 200, "/api/db/base/entitiesByName/{name}", "");
                });
    }
}

File: e2e-tests/spring-graphql/src/test/kotlin/org/evomaster/e2etests/spring/graphql/db/base/DbBaseEMTest.kt
package org.evomaster.e2etests.spring.graphql.db.base

import com.foo.graphql.db.base.DbBaseController
import org.evomaster.core.EMConfig
import org.evomaster.core.problem.graphql.GQMethodType
import org.evomaster.e2etests.spring.graphql.SpringTestBase
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import java.util.*
import kotlin.booleanArrayOf

class DbBaseEMTest : SpringTestBase() {


    companion object {
        @BeforeAll
        @JvmStatic
        fun init() {
            initClass(DbBaseController())
        }
    }

    @ParameterizedTest
    @ValueSource(booleans = [false, true])
    fun testRunEM(heuristicsForSQLAdvanced: Boolean) {
        runTestHandlingFlakyAndCompilation(
            "DbBaseEM" +
                    if (heuristicsForSQLAdvanced) "Complete" else "Partial",
            10000
        ) { args: MutableList<String> ->

            /*
                TODO
                Known issue with string escaping in generated tests, eg \u.
                until fixed, this test might fail, and need to change its seed
             */
            defaultSeed = 3

            setOption(args, "problemType", EMConfig.ProblemType.GRAPHQL.toString())
            setOption(args, "heuristicsForSQL", "true")
            setOption(args, "generateSqlDataWithSearch", "false")
            setOption(args, "heuristicsForSQLAdvanced", if (heuristicsForSQLAdvanced) "true" else "false")

            val solution = initAndRun(args)

            assertTrue(solution.individuals.size >= 1)
            assertHasAtLeastOne(
                solution,
                "dbBaseByName",
                GQMethodType.QUERY,
                200,
                Arrays.asList("\"id\":\"42\"", "\"name\":\"foo\""),
                false
            )
            // there exists some problems on addDbBase, e.g., 500 MUTATION addDbBase, auth=NoAuth
            // assertNoneWithErrors(solution)
        }
    }
}
